### DISCUSSION TITLE
nested relationships and duplication function

### DISCUSSION THREAD

**Discussion by Dumpstar56** [directus/directus#14635] [Archive]
Hi everyone,
I am losing my mind with the duplication function. I know I only need to select all the fields which should be copied, but once it gets to relationships, some nested 2 or 3 tables deep, I cannot figure it out. Here are two cases:
I have created a test setup of a parent, o2m child, and m2o child's child table. Each has test string or number values. The copy is supposed to create a new item in the parent and the child tables but refer to the same item in the child's child table. I have checked these options for the duplication fields.
![duplication1](https://user-images.githubusercontent.com/94397183/180815586-20ad0caa-35cb-46b1-b32c-21f323eac37c.jpg)

However, when I use save as copy, some edits cause the copied item to not include the values in the child table. They appear as blank in the child, even thought the nested relationship to the child's child item copies fine. Is there something wrong with my settings?

In my actual use case, the tables, values, and relationships are even more complex and nested deeper, most often with an additional m2m from the child table, and a parent's parent table. There, I only get a spinny wheel when I try to save as copy, even though I have essentially set up the fields to copy like my tester.

If I have the parent's foreign ID field in the child table and the copied item should retain the relationship to the newly copied parent, rather than the original parent, I still select that foreign ID to be copied, correct? Do I also select it, if the parent is not being copied and the only copies are the child records? What happens if I do not select the foreign ID field? Does it lose the relationship, or does it stay related to the original record?

Does anybody have clarifications on the duplication function and nested relationships? Which fields do I select for what to happen? And how do I avoid the function failing?

Thank you so much in advance!

**Comment by rijkvanzanten** [comment]
> However, when I use save as copy, some edits cause the copied item to not include the values in the child table. 

What do you mean by _some edits_? When the user has already made other edits to the relational field you mean?

> There, I only get a spinny wheel when I try to save as copy,

Any errors in the dev tools / network tab? This to me sounds like the API is timing out or throwing an error of sorts.

> If I have the parent's foreign ID field in the child table and the copied item should retain the relationship to the newly copied parent, rather than the original parent, I still select that foreign ID to be copied, correct? 

This is the confusing bit. If you're talking about a one-to-many, the nested `parent_id` contains the primary key of the current "parent". If you are duplicating the current parent, you will create a _new_ item with a new primary key, and should therefore not copy the old `parent_id` (as that points to the thing you copied, not the new thing that saved). 

> What happens if I do not select the foreign ID field? Does it lose the relationship, or does it stay related to the original record?

The API will automatically set the parent_id field to the currently created item if it's undefined, as it's created nested through the o2m. (For example, posting `{ o2m_field: [{ value: 'test' }] }}` to the API will set the `parent_id` of the newly created thing to the parent item).

> Does anybody have clarifications on the duplication function and nested relationships? Which fields do I select for what to happen? 

The way the "save as copy" works is that the app will first re-fetch the current item from the API using the above configuration, then apply the edits the user already made on top of that, and POST that to the API to create a new item. This also means that you can effectively treat this config as a "read". With that in mind, all the regular API rules around inserting nested related data comes into play. Submitting `{ m2o_field: { text: 'value' }}` will create a _new_ related m2o item with text: value. Submitting `{ m2o_field: { id: 1 }}` will set the `m2o_field` to `1`, which makes it the same as `m2o_field: 1`. https://docs.directus.io/reference/introduction/#relational-data

Once you get into multi-nested relational data, I agree that the configuration becomes a bit confusing!

**Comment by Dumpstar56** [comment]
Thank you so much for the long answer to my long question. I deselected the parent ID field in the child table so it will be related to the newly copied item. However, if I edit the value in the Child Table - Text in Child above - I get the error below, but the item in the aprent table copies with a newly created child item that only has the edited valie in Text in Child but Text2 in Child and the relation to the nested m2o child's child table are null. Am I doing something wrong?

```json
{
  "errors": [
    {
      "message": "select \"test_duplo2m\".\"id\" from \"test_duplo2m\" where \"test_duplo2m\".\"parent_id\" = $1 order by \"test_duplo2m\".\"id\" asc limit $2 - invalid input syntax for integer: \"NaN\"",
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ]
}
```

**Reply by rijkvanzanten** [reply]
I don't fully understands the steps you took to get to this error. If possible, would you be able to share a database dump with a minimal reproduction we can use to take a look and try it out too? 🙂

**Reply by Dumpstar56** [reply]
Access to the db in our case is complicated, so I could try for a dump tomorrow. Sorry.
For now I can give you this step by step description of what I did, which might help.

1. Create table with text field 1, text field 2, and a o2m relation to a new table called test_o2m
2. In test_o2m create text field 1, text field 2, and a m2o relation to a new table called test_m2o
3. In test_m2o, create number field 1 (text should also work, but I created a number field)
4. Set copy function like this:
![copyerror](https://user-images.githubusercontent.com/94397183/181635232-9d9a495b-0e27-4d37-8e9c-cb1d981ebfb3.jpg)
Please let me know if you would still need a db dump.

**Reply by rijkvanzanten** [reply]
Thanks! And what database type is this on? (That sometimes affects this type of thing)

**Reply by Dumpstar56** [reply]
Which version of Node.js are you using?
Which database are you using? Postgresql 14?
How are you deploying Directus?

- Version of Node.js: v14.18.1
- Version of Postgres: 10.14 (AWS Aurora is always behind, though we may be able to upgrade to 13)
- AWS EC2, pm2 for app management, nginx reverse proxy.

**Reply by rijkvanzanten** [reply]
@licitdev mind helping debug this? Wondering if it's somehow related to https://github.com/directus/directus/issues/13107

**Comment by licitdev** [comment]
@Dumpstar56 Which version of Directus are you using? The `NaN` error seems to be similar to https://github.com/directus/directus/issues/14209 which has been resolved.

### Test Findings
New items are created for the parent & o2m child tables with the correct values with the nested m2o being referenced. 

https://user-images.githubusercontent.com/26413686/181729423-fa0b1aaa-ac45-44d9-9f8e-60272c265b0d.mov

**Reply by Dumpstar56** [reply]
I am using Directus 9.14.5.
Try editing text in child 1 or 2 and then save as copy. That is when it happens to me.

**Reply by licitdev** [reply]
I can replicate the issue now. I've filed an issue https://github.com/directus/directus/issues/14744.
Thanks!
